empress/recon_vis/recon_viewer.py

```diff
@@ -1,80 +1,206 @@
+    tip_font_size, internal_font_size = calculate_font_size(num_tips, num_nodes)
 
-    fig = plot_tools.FigureWrapper(consistency_type, axes)
-    if consistency_type != utils.ConsistencyType.NO_CONSISTENCY:
-        render_host(fig, host_tree, show_internal_labels)
-        host_lookup = host_tree.name_to_node_dict()
-        render_parasite(fig, parasite_tree, recon, host_lookup, show_internal_labels, show_freq)
-    return fig
+    root = parasite_tree.root_node
+    host_lookup = host_tree.name_to_node_dict()
+    parasite_lookup = parasite_tree.name_to_node_dict()
+    
+    #Populate Host Nodes with track count
+    populate_host_tracks(root, recon, host_lookup)
+    
+    #Render Host Tree
+    render_host(fig, host_tree, show_internal_labels, tip_font_size, internal_font_size)
+    
+    #Sets the offsets between tracks on each host node
+    set_offsets(host_tree)
+    
+    #Render Parasite Tree
+    render_parasite(fig, parasite_tree, recon, host_lookup, parasite_lookup, show_internal_labels, show_freq, tip_font_size, internal_font_size)
+
+    return fig 
 
+def create_legend(fig, consistency_type):
+    legend_elements = [
+                        Line2D([0], [0], marker= COSPECIATION_NODE_SHAPE, color='w', label='Cospeciation', \
+                               markerfacecolor=COSPECIATION_NODE_COLOR, markersize=NODESIZE),
+                        Line2D([0], [0], marker=DUPLICATION_NODE_SHAPE, color='w', label='Duplication', \
+                               markerfacecolor=DUPLICATION_NODE_COLOR, markersize=NODESIZE),
+                        Line2D([0], [0], marker=TRANSFER_NODE_SHAPE, color='w', label='Transfer', \
+                               markerfacecolor=TRANSFER_NODE_COLOR, markersize=NODESIZE),\
+                        LineCollection( [[(0, 0)]], linestyles = ['dashed'], \
+                                       colors = [LOSS_EDGE_COLOR], label='Loss')
+                      ]
+    fig.set_legend(legend_elements, title = consistency_type)
+def set_offsets(tree):
+    """
+    Populates the nodes of a Tree with an offset
+    :param tree: Tree Object
+    """
+    pos_dict = tree.pos_dict
+    
+    for node in tree.postorder_list:
+        y_1 = None
+        y_0 = node.layout.row
+        for logical_pos in pos_dict:
+            if node.is_leaf:
+                if y_0 < logical_pos[0] and node.layout.col <= logical_pos[1]:
+                    if y_1 == None or y_1 > logical_pos[0]:
+                        y_1 = logical_pos[0]
+            else:
+                y_1 = max(node.left_node.layout.row, node.right_node.layout.row)
+        if y_1 == None or node.layout.node_count == 0:
+            node.layout.offset = TRACK_OFFSET
+        else:
+            node.layout.offset = abs(y_0 - y_1) / (node.layout.node_count + 3)
 
-def render_host(fig, host_tree: tree.Tree, show_internal_labels):
-    """ Renders the host tree """
+def render_host(fig, host_tree, show_internal_labels, tip_font_size, internal_font_size):
```

Comment:
It's okay if you don't want to add type hints to your code, but please don't remove the existing type hints.

Reply:
I should mention that we copied and pasted our entire recon_viewer.py file over the old one. We have no idea what changed design wise. Would you like us to have type hints in our function headers?

Reply:
It's okay if you don't want to add type hints or comments to codes that don't interact with other parts of the codebase. Type hints would be nice, but it's not necessary. The decision is totally up to you.

Reply:
My understanding is that `render_host` does not interact with other parts of the code? If this the case, I will retract what I said earlier about " please don't remove the existing type hints". If you decide to remove type hints, please remove all of them from private codes.

Reply:
Commenting "Done" so I can keep track of what has been fixed already.